[FSDP] Short-term fix to remove `optim_input`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#84201 [FSDP] Short-term fix to remove `optim_input`**
* #83917 [FSDP] Add rate limiter
* #84366 [FSDP][Easy] Save unpadded/padded unsharded sizes as attributes
* #83665 [FSDP] Generalize prefetching; lower unshard/reshard to handle
* #84761 [FSDP][Easy] Minor cleanup
* #84601 [FSDP] Subtest prefetching for `test_fsdp_grad_acc.py`
* #84600 [FSDP] Remove `forward_prefetch`

This is a short-term quick fix to accommodate using the existing optimizer state APIs without passing `optim_input`. It preserves the existing `optim_input` code path but if `optim_input` is `None` while `optim` is not, then the APIs will use the new code path that relies on `self.param_groups` to get the information previously provided by `optim_input`.

Commit:
[FSDP] Short-term fix to remove `optim_input`

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -858,31 +858,26 @@ def _broadcast_unsharded_pos_dim_tensor_state(
-    which may be different across ranks. In particular, the unflattened
-    parameter names are represented as :class:`_OptimStateKey` s.
+    flattened parameter IDs according to the calling rank's ``optim``, which
+    may be different across ranks. In particular, the unflattener parameter
```
don't quite understand "unflattener parameter"

awgu replied:
```suggestion
    may be different across ranks. In particular, the unflattened parameter
```

When I first onboarded onto FSDP, I had a mental of **flattened parameters** (`FlatParameter`s) and **unflattened parameters** (original parameters), so all of the optimizer state dict work uses that language. I think now we have collectively drifted to using the term **original parameters** instead of **unflattened parameters**.

I think **original parameters** and then **FSDP parameters** might be solid language, where **FSDP parameters** generalizes (`FlatParameter`s) in case we ever choose to combine original parameters in a non-flattened way. However, I am certainly open to discussion.

kumpera replied:
flattened/unflattened, sharded/unsharded, those are terms that usually don't mean much to external users.

awgu replied:
Mhm, makes sense.

I was referring to within the FSDP devs. Since we have several threads going on right now using terms like flattened/unflattened, sharded/unsharded, original/FSDP etc., it might be helpful to converge to some common terminology.

rohan-varma commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -858,31 +858,26 @@ def _broadcast_unsharded_pos_dim_tensor_state(
-            Iterable[torch.nn.Parameter],
-        ]
-    ] = None,
+    optim: torch.optim.Optimizer,
```
one suggestion for backwards compatibility would be to maintain the old codepath, add this new argument as optional, and log a warning indicating to use this argument and mention that the old one will be deprecated.

Although, since FSDP is beta and doesn't guarantee BC yet, this change + fixing the callsites we know of should be acceptable as well.

kumpera replied:
This is an internal API, it's ok to change it at this point.

awgu replied:
I followed @rohan-varma and @fegin's suggestion to preserve the existing code path so that we do not need to migrate internal customers in the short term. We are playing to revamp the API soon anyway, so to move faster, it makes sense to preserve the `optim_input` code path for the customers using it. The goal after this PR is to accommodate users that want to use the existing APIs (before we revamp) but do not want to pass `optim_input`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3721,9 +3702,7 @@ def sharded_optim_state_dict(
-        optim_input: Optional[Union[
-            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
-        ]] = None,
+        optim: torch.optim.Optimizer,
```
curious why we're getting rid of `None` default codepath and enforcing the arg?

kumpera replied:
Similarly, do we need `full_optim_state_dict` when it's just `optim.state_dict()`??

awgu replied:
I do not believe that `full_optim_state_dict()` is the same as `optim.state_dict()`.

Each `optim` is a per-rank local optimizer that operates in terms of sharded `FlatParameter`s. The return value of `full_optim_state_dict()` is in terms of unsharded original parameters.

rohan-varma commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -858,31 +858,26 @@ def _broadcast_unsharded_pos_dim_tensor_state(
-            Iterable[torch.nn.Parameter],
-        ]
-    ] = None,
+    optim: torch.optim.Optimizer,
```
the previous code used to allow this not to be specified and by default use `list(model.parameters())`, is this something that we'd like to maintain?

awgu replied:
Resolved by preserving the existing code path.

rohan-varma reviewed (commented):
Looks good, just a couple of questions for your consideration and will stamp after that.

fegin commented:
Have the same comment as @rohan-varma , do we want to keep the backward compatibility? We can have the `optim_state_dict` for users who are eager to get rid of `optim_input` without touching the existing APIs. Could this be faster and safer?

Commit:
Update on "[FSDP] Short-term fix to remove `optim_input`"


This is a short-term quick fix to retire the `optim_input` argument. It directly use the `self.param_groups` attribute instead.

This change requires additional internal changes to training code to be backward compatible since it removes the `optim_input` argument.

[ghstack-poisoned]

awgu commented:
@rohan-varma @fegin I updated the PR to preserve the old code path. If `optim` is `None`, then it takes the existing `optim_input` code path. If `optim_input` is `None` and `optim` is not `None`, then it takes the new code path.

Commit:
Update on "[FSDP] Short-term fix to remove `optim_input`"


This is a short-term quick fix to retire the `optim_input` argument. It directly use the `self.param_groups` attribute instead.

This change requires additional internal changes to training code to be backward compatible since it removes the `optim_input` argument.

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -473,6 +473,23 @@ def test_optim_state_dict_nested(
         are incorrectly mapped to values. Their correct mapping is tested in
         other tests that exercise the save/load workflow.
         """
+        self.run_subtests(
```
All but a few edge case tests now subtest both `use_optim_input=True` and `False` to test both code paths.

rohan-varma replied:
awesome!

Commit:
Update on "[FSDP] Short-term fix to remove `optim_input`"


This is a short-term quick fix to accommodate using the existing optimizer state APIs without passing `optim_input`. It preserves the existing `optim_input` code path but if `optim_input` is `None` while `optim` is not, then the APIs will use the new code path that relies on `self.param_groups` to get the information previously provided by `optim_input`.


[ghstack-poisoned]

Commit:
Update on "[FSDP] Short-term fix to remove `optim_input`"


This is a short-term quick fix to accommodate using the existing optimizer state APIs without passing `optim_input`. It preserves the existing `optim_input` code path but if `optim_input` is `None` while `optim` is not, then the APIs will use the new code path that relies on `self.param_groups` to get the information previously provided by `optim_input`.


[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM overall, thanks for taking the suggestion to maintain BC for now with the old code path! Please ensure @kumpera's comments are resolved as well or let's discuss offline if additional clarification is needed.

Commit:
Update on "[FSDP] Short-term fix to remove `optim_input`"


This is a short-term quick fix to accommodate using the existing optimizer state APIs without passing `optim_input`. It preserves the existing `optim_input` code path but if `optim_input` is `None` while `optim` is not, then the APIs will use the new code path that relies on `self.param_groups` to get the information previously provided by `optim_input`.


[ghstack-poisoned]

awgu commented:
@pytorchbot merge
